torch/distributed/device_mesh.py

```diff
@@ -1159,6 +1159,52 @@ def _unflatten(
                 self, dim, mesh_sizes, mesh_dim_names, backend_override
             )
 
+        @staticmethod
+        def _concatenate(device_mesh_list: list["DeviceMesh"]) -> "DeviceMesh":
+            concatenate_dim_names: list[str] = []
+            concatenate_sizes: list[int] = []
+            concatenate_strides: list[int] = []
+            concatenate_dim_group_name: list[str] = []
+            root_mesh = _mesh_resources.get_root_mesh(device_mesh_list[0])
+            for mesh in device_mesh_list:
+                concatenate_dim_names.extend(not_none(mesh.mesh_dim_names))
+                concatenate_sizes.extend(list(flatten(mesh._layout.sizes)))
+                concatenate_strides.extend(list(flatten(mesh._layout.strides)))
+                concatenate_dim_group_name.extend(not_none(mesh._dim_group_names))
+                # Concatenate device mesh having different root mesh tensors are meaningless
+                # because the concatenated indices should be indexed by the same root mesh tensor.
+                if not torch.equal(
+                    root_mesh.mesh, _mesh_resources.get_root_mesh(mesh).mesh
+                ):
+                    raise RuntimeError(
+                        "Cannot concatenate DeviceMeshes with different root mesh tensors"
+                    )
+            concatenate_mesh_layout = _MeshLayout(
+                tuple(concatenate_sizes), tuple(concatenate_strides)
+            )
+            if not concatenate_mesh_layout.check_non_overlap():
+                raise RuntimeError(
+                    f"Cannot concatenate overlapping meshes: {device_mesh_list}"
```

Comment:
I do wonder if someone is going to ask for a concatenate that implicitly drops duplicated dims. This would be coherent I think, though I don't know if you need it.